File: typescript/docs/examples/browser-use/browser/browser-type-example.ts
// @ts-nocheck
/**
 * Browser Type Selection Example
 *
 * This example demonstrates how to select between Chrome and Chromium browsers
 * when using computer use images in AgentBay.
 *
 * Features demonstrated:
 * - Chrome browser selection
 * - Chromium browser selection
 * - Default browser (undefined)
 * - Browser type verification
 * - Configuration comparison
 *
 * Note: The browserType option is only available for computer use images.
 */

import { AgentBay, CreateSessionParams, BrowserOption } from 'wuying-agentbay-sdk';
import { chromium } from 'playwright';

/**
 * Test a specific browser type configuration
 */
async function testBrowserType(
  browserType: 'chrome' | 'chromium' | undefined,
  description: string
): Promise<void> {
  console.log('\n' + '='.repeat(60));
  console.log(`Testing: ${description}`);
  console.log('='.repeat(60));

  const apiKey = process.env.AGENTBAY_API_KEY;
  if (!apiKey) {
    throw new Error('AGENTBAY_API_KEY environment variable not set');
  }

  const agentBay = new AgentBay({ apiKey });

  // Create session with computer use image
  console.log('\n1. Creating session with computer use image...');
  const params: CreateSessionParams = { imageId: 'browser_latest' };
  const result = await agentBay.create(params);

  if (!result.success || !result.session) {
    throw new Error('Failed to create session');
  }

  const session = result.session;
  console.log(`   ✓ Session created: ${session.getSessionId()}`);

  try {
    // Initialize browser with specified type
    console.log(`\n2. Initializing browser with type: ${browserType || 'default (undefined)'}...`);
    
    // Using plain object (recommended)
    const option: BrowserOption = {
      browserType: browserType,
      viewport: { width: 1920, height: 1080 }
    };

    const success = await session.browser.initializeAsync(option);
    if (!success) {
      throw new Error('Browser initialization failed');
    }

    console.log('   ✓ Browser initialized successfully');

    // Get endpoint URL
    const endpointUrl = session.browser.getEndpointUrl();
    console.log(`\n3. CDP endpoint: ${String(endpointUrl).substring(0, 50)}...`);

    // Connect Playwright and verify browser
    console.log('\n4. Connecting to browser via CDP...');
    const browser = await chromium.connectOverCDP(endpointUrl);
    const context = browser.contexts()[0];
    const page = await context.newPage();

    console.log('   ✓ Connected successfully');
    console.log('\n5. Verifying browser configuration...');

    // Navigate to a page that shows browser info
    await page.goto('https://www.whatismybrowser.com/');
    await page.waitForLoadState('networkidle');

    // Get browser information
    const userAgent = await page.evaluate(() => (navigator as any).userAgent);
    const viewportWidth = await page.evaluate(() => (window as any).innerWidth);
    const viewportHeight = await page.evaluate(() => (window as any).innerHeight);

    console.log('\n   Browser Information:');
    console.log(`   - User Agent: ${userAgent.substring(0, 80)}...`);
    console.log(`   - Viewport: ${viewportWidth} x ${viewportHeight}`);
    console.log(`   - Configured Type: ${browserType || 'default'}`);

    // Check if Chrome or Chromium is in user agent
    if (userAgent.includes('Chrome')) {
      const detected = userAgent.includes('Chromium') ? 'Chromium' : 'Chrome';
      console.log(`   - Detected Browser: ${detected}`);
    }

    await browser.close();

    console.log(`\n   ✓ Test completed successfully for ${description}`);

  } finally {
    console.log('\n6. Cleaning up...');
    session.delete();
    console.log('   ✓ Session deleted');
  }
}

/**
 * Main function to run all browser type examples
 */
async function main(): Promise<void> {
  console.log('Browser Type Selection Example');
  console.log('='.repeat(60));
  console.log('\nThis example demonstrates browser type selection in AgentBay.');
  console.log('Note: browserType is only available for computer use images.');

  // Test 1: Chrome browser
  await testBrowserType('chrome', 'Chrome Browser (Google Chrome)');

  await new Promise(resolve => setTimeout(resolve, 2000)); // Brief pause

  // Test 2: Chromium browser
  await testBrowserType('chromium', 'Chromium Browser (Open Source)');

  await new Promise(resolve => setTimeout(resolve, 2000)); // Brief pause

  // Test 3: Default (undefined)
  await testBrowserType(undefined, 'Default Browser (Platform decides)');

  console.log('\n' + '='.repeat(60));
  console.log('All browser type tests completed successfully!');
  console.log('='.repeat(60));

  // Summary
  console.log('\nSummary:');
  console.log('- Chrome: Use when you need Google Chrome specific features');
  console.log('- Chromium: Use for open-source, lighter resource usage');
  console.log('- Default (undefined): Let the platform choose the optimal browser');
  console.log('\nBest Practice: Use undefined unless you have a specific requirement');
}

/**
 * Quick example showing the most common usage
 */
async function quickExample(): Promise<void> {
  console.log('\n' + '='.repeat(60));
  console.log('Quick Example: Using Chrome Browser');
  console.log('='.repeat(60));

  const apiKey = process.env.AGENTBAY_API_KEY;
  if (!apiKey) {
    throw new Error('AGENTBAY_API_KEY not set');
  }

  const agentBay = new AgentBay({ apiKey });

  // Create session
  const params: CreateSessionParams = { imageId: 'browser_latest' };
  const result = await agentBay.create(params);
  if (!result.success || !result.session) {
    throw new Error('Failed to create session');
  }

  const session = result.session;

  try {
    // Simply specify browserType in BrowserOption
    const option: BrowserOption = {
      browserType: 'chrome'
    };

    const success = await session.browser.initializeAsync(option);

    if (success) {
      console.log('✓ Chrome browser initialized successfully');

      // Get endpoint and use with Playwright
      const endpointUrl = session.browser.getEndpointUrl();
      const browser = await chromium.connectOverCDP(endpointUrl);
      const page = await browser.contexts()[0].newPage();

      await page.goto('https://example.com');
      const title = await page.title();
      console.log(`✓ Page title: ${title}`);

      await browser.close();
    }
  } finally {
    session.delete();
  }
}

/**
 * Type-safe example showing all options
 */
async function typeSafeExample(): Promise<void> {
  console.log('\n' + '='.repeat(60));
  console.log('Type-Safe Example: Using TypeScript Types');
  console.log('='.repeat(60));

  const apiKey = process.env.AGENTBAY_API_KEY;
  if (!apiKey) {
    throw new Error('AGENTBAY_API_KEY not set');
  }

  const agentBay = new AgentBay({ apiKey });
  const params: CreateSessionParams = { imageId: 'browser_latest' };
  const result = await agentBay.create(params);

  if (!result.success || !result.session) {
    throw new Error('Failed to create session');
  }

  const session = result.session;

  try {
    // TypeScript ensures type safety
    const browserTypes: Array<'chrome' | 'chromium' | undefined> = [
      'chrome',
      'chromium',
      undefined
    ];

    for (const type of browserTypes) {
      console.log(`\nTesting with type: ${type || 'undefined'}`);
      
      // Type-safe option creation
      const option: BrowserOption = {
        browserType: type,
        viewport: { width: 1920, height: 1080 }
      };

      const success = await session.browser.initializeAsync(option);
      console.log(`  ✓ Initialization: ${success ? 'Success' : 'Failed'}`);

      if (success) {
        // Verify configuration
        const currentOption = session.browser.getOption();
        if (currentOption) {
          console.log(`  ✓ Configured type: ${currentOption.browserType || 'undefined'}`);
        }
      }

      // Brief pause between iterations
      await new Promise(resolve => setTimeout(resolve, 1000));
    }

    console.log('\n✓ Type-safe example completed');

  } finally {
    session.delete();
  }
}

// Run the main example
main()
  .then(() => process.exit(0))
  .catch(error => {
    console.error('Error:', error.message);
    process.exit(1);
  });

File: python/docs/examples/browser-use/browser/browser_type_example.py
"""
Browser Type Selection Example

This example demonstrates how to select between Chrome and Chromium browsers
when using computer use images in AgentBay.

Features demonstrated:
- Chrome browser selection
- Chromium browser selection
- Default browser (None)
- Browser type verification
- Configuration comparison

Note: The browser_type option is only available for computer use images.
"""

import os
import asyncio
from agentbay import AgentBay
from agentbay.session_params import CreateSessionParams
from agentbay.browser.browser import BrowserOption, BrowserViewport
from playwright.async_api import async_playwright


async def test_browser_type(browser_type: str | None, description: str):
    """Test a specific browser type configuration."""
    print(f"\n{'='*60}")
    print(f"Testing: {description}")
    print(f"{'='*60}")

    api_key = os.getenv("AGENTBAY_API_KEY")
    if not api_key:
        raise RuntimeError("AGENTBAY_API_KEY environment variable not set")

    agent_bay = AgentBay(api_key=api_key)

    # Create session with computer use image
    print("\n1. Creating session with computer use image...")
    params = CreateSessionParams(image_id="browser_latest")
    result = agent_bay.create(params)

    if not result.success:
        raise RuntimeError(f"Failed to create session: {result.error_message}")

    session = result.session
    print(f"   ✓ Session created: {session.session_id}")

    try:
        # Initialize browser with specified type
        print(f"\n2. Initializing browser with type: {browser_type or 'default (None)'}")
        option = BrowserOption(
            browser_type=browser_type,
            viewport=BrowserViewport(width=1920, height=1080)
        )

        success = await session.browser.initialize_async(option)
        if not success:
            raise RuntimeError("Browser initialization failed")

        print(f"   ✓ Browser initialized successfully")

        # Get endpoint URL
        endpoint_url = session.browser.get_endpoint_url()
        print(f"\n3. CDP endpoint: {endpoint_url[:50]}...")

        # Connect Playwright and verify browser
        print("\n4. Connecting to browser via CDP...")
        async with async_playwright() as p:
            browser = await p.chromium.connect_over_cdp(endpoint_url)
            context = browser.contexts[0]
            page = await context.new_page()

            # Navigate to a page that shows browser info
            print("   ✓ Connected successfully")
            print("\n5. Verifying browser configuration...")

            await page.goto("https://www.whatismybrowser.com/")
            await page.wait_for_load_state("networkidle")

            # Get browser information
            user_agent = await page.evaluate("navigator.userAgent")
            viewport_width = await page.evaluate("window.innerWidth")
            viewport_height = await page.evaluate("window.innerHeight")

            print(f"\n   Browser Information:")
            print(f"   - User Agent: {user_agent[:80]}...")
            print(f"   - Viewport: {viewport_width} x {viewport_height}")
            print(f"   - Configured Type: {browser_type or 'default'}")

            # Check if Chrome or Chromium is in user agent
            if "Chrome" in user_agent:
                if "Chromium" in user_agent:
                    detected = "Chromium"
                else:
                    detected = "Chrome"
                print(f"   - Detected Browser: {detected}")

            await browser.close()

        print(f"\n   ✓ Test completed successfully for {description}")

    finally:
        print("\n6. Cleaning up...")
        session.delete()
        print("   ✓ Session deleted")


async def main():
    """Run browser type examples."""
    print("Browser Type Selection Example")
    print("=" * 60)
    print("\nThis example demonstrates browser type selection in AgentBay.")
    print("Note: browser_type is only available for computer use images.")

    # Test 1: Chrome browser
    await test_browser_type(
        browser_type="chrome",
        description="Chrome Browser (Google Chrome)"
    )

    await asyncio.sleep(2)  # Brief pause between tests

    # Test 2: Chromium browser
    await test_browser_type(
        browser_type="chromium",
        description="Chromium Browser (Open Source)"
    )

    await asyncio.sleep(2)  # Brief pause between tests

    # Test 3: Default (None)
    await test_browser_type(
        browser_type=None,
        description="Default Browser (Platform decides)"
    )

    print("\n" + "=" * 60)
    print("All browser type tests completed successfully!")
    print("=" * 60)

    # Summary
    print("\nSummary:")
    print("- Chrome: Use when you need Google Chrome specific features")
    print("- Chromium: Use for open-source, lighter resource usage")
    print("- Default (None): Let the platform choose the optimal browser")
    print("\nBest Practice: Use None unless you have a specific requirement")


async def quick_example():
    """Quick example showing the most common usage."""
    print("\n" + "=" * 60)
    print("Quick Example: Using Chrome Browser")
    print("=" * 60)

    api_key = os.getenv("AGENTBAY_API_KEY")
    agent_bay = AgentBay(api_key=api_key)

    # Create session
    params = CreateSessionParams(image_id="browser_latest")
    result = agent_bay.create(params)
    session = result.session

    try:
        # Simply specify browser_type in BrowserOption
        option = BrowserOption(browser_type="chrome")
        success = await session.browser.initialize_async(option)

        if success:
            print("✓ Chrome browser initialized successfully")

            # Get endpoint and use with Playwright
            endpoint_url = session.browser.get_endpoint_url()
            async with async_playwright() as p:
                browser = await p.chromium.connect_over_cdp(endpoint_url)
                page = await browser.contexts[0].new_page()

                await page.goto("https://example.com")
                title = await page.title()
                print(f"✓ Page title: {title}")

                await browser.close()
    finally:
        session.delete()


if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser(description="Browser Type Selection Example")
    parser.add_argument(
        "--quick",
        action="store_true",
        help="Run quick example only"
    )

    args = parser.parse_args()

    if args.quick:
        asyncio.run(quick_example())
    else:
        asyncio.run(main())